### PR TITLE
Allow JWT::decode to accept an empty string as a valid kid

### PR DESCRIPTION
**Pull request by ryanneufeld**
There are instances when using CachedKeySet where a key is returned with an empty string as the kid. This is a valid use case and should be allowed.

For example Teleport Proxy uses this pattern to allow for a default key.
```
{
    "keys": [
        {
            "kty": "RSA",
            "alg": "RS256",
            "n": "<redacted>",
            "e": "AQAB",
            "use": "sig",
            "kid": ""
        }
    ]
}
```
Newer versions of teleport also return:

```
{
    "keys": [
        {
            "kty": "RSA",
            "alg": "RS256",
            "n": "<same key redacted>",
            "e": "AQAB",
            "use": "sig",
            "kid": "<redacted kid>"
        },
        {
            "kty": "RSA",
            "alg": "RS256",
            "n": "<same key redacted>",
            "e": "AQAB",
            "use": "sig",
            "kid": ""
        }
    ]
}
```

The getKey method can be simplified, as well as refactored to follow the same pattern as the CachedKeySet class which casts null kids to an empty string.

This change also adds a test to ensure that an empty string kid is a valid kid.

### CODE REVIEW

**Comment by ryanneufeld** [comment]
I've signed the CCA, is there anyone who has time to review this?

**Comment by tabacco** [comment]
It's not explicitly specified, except that it's a case-sensitive string: https://datatracker.ietf.org/doc/html/rfc7517#section-4.5

In the wild, I've observed JWK data with an extra key containing a kid of "" from [Teleport](https://goteleport.com/), so it's definitely a thing in the wild.

**Comment by ryanneufeld** [comment]
> @ryanneufeld thank you for your submission! I'll take a look.
> 
> Do you know if the spec defines this as a valid use of `kid`? Or is it primarily that you've observed this behavior in the wild as a valid use case?

**Comment by ryanneufeld** [comment]
> @ryanneufeld thank you for your submission! I'll take a look.
> 
> Do you know if the spec defines this as a valid use of `kid`? Or is it primarily that you've observed this behavior in the wild as a valid use case?

As Doug mentioned. There are use-cases in the wild where this patch would enable support without special code to handle it at the implementation layer.

**Comment by ryanneufeld** [comment]
@bshaffer Any updates on this?

File: src/JWT.php [hunk 1]
```diff
@@ -460,17 +460,15 @@ private static function getKey(
...
             return $keyOrKeyArray;
         }
 
-        if (empty($kid) && $kid !== '0') {
-            throw new UnexpectedValueException('"kid" empty, unable to lookup correct key');
-        }
-
-        if ($keyOrKeyArray instanceof CachedKeySet) {
-            // Skip "isset" check, as this will automatically refresh if not set
-            return $keyOrKeyArray[$kid];
```
Comment (bshaffer) [inline src/JWT.php:473]: Why are you removing these lines?

**Review by bshaffer — COMMENTED** [review]
It's not clear to me this functionality is a valid use of the spec. For that reason I'm reticent to add it.

File: src/JWT.php (hunk 1 above)
Comment (ryanneufeld) [reply src/JWT.php:473]: The new lines below cover this test as the CachedKeySet implements ArrayAccess

The resulting logic is the same 
https://github.com/firebase/php-jwt/blob/main/src/CachedKeySet.php#L19C1-L19C42

**Comment by bshaffer** [comment]
@ryanneufeld are you being serious?

**Comment by ryanneufeld** [comment]
> @ryanneufeld are you being serious?

I'm not sure how to respond to this question, you said you're not sure if it meets the spec. I'm not sure which spec you're referring to, so if you were to point me in the right direction I could address your concerns.

So, yes, I am being serious, not combative.

File: src/JWT.php (hunk 1 above)
Comment (bshaffer) [reply src/JWT.php:473]: If you read the comment of the removed line, you'll see that the `isset` check triggers a refresh of the credentials, which we don't want to or need to do.

**Comment by bshaffer** [comment]
My concern is not that the spec forbids it, my concern is that this adds logic which is not defined by any spec.

Supporting an empty string as a KID is one thing, but selecting a key by default when the KID is empty is another.

**Comment by ryanneufeld** [comment]
@bshaffer I see your point. I suppose the idea here is that we need *a* way to pick a key.

The idea here is that if there are two keys, and has an empty KID, and no KID is passed in for selection, then allow the one with the empty key to match.

The alternative involves parsing the JWT's multiple times in order to validate them. 

I welcome any suggestions/collaboration you have to offer for resolving this situation.

**Comment by ryanneufeld** [comment]
@bshaffer After looking into more closely, I made a refactor that I think will satisfy your concerns, while maintaining the previous functionality. 

Additionally this will increase the code coverage with some more tests.
